tests/unit/test_kdf.cpp

```diff
@@ -0,0 +1,201 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include "util.cpp"
+
+#include "mococrw/kdf.h"
+#include "mococrw/hash.h"
+
+using namespace mococrw;
+
+using testing::Eq;
+
+struct inputData
+{
+    openssl::DigestTypes hashFunction;
+    std::vector<uint8_t> password;
+    std::vector<uint8_t> salt;
+    uint iterations;
+    uint olen;
+    std::string expectedRestult;
+};
+
+void testPbkfd2(struct inputData data)
+{
+    auto pbkfd2 = PBKDF2(data.hashFunction, data.iterations);
+    ASSERT_THAT(utility::toHex(pbkfd2.deriveKey(data.password, data.olen, data.salt)),
+                Eq(data.expectedRestult));
+}
+
+static std::vector<inputData> prepareTestDataForPbkdf2Tests()
+{
+    std::vector<inputData> testData {
+        {
+            openssl::DigestTypes::SHA1,
+            {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+            {'s', 'a', 'l', 't'},
+            1,
+            20,
+            "0c60c80f961f0e71f3a9b524af6012062fe037a6"
+        },
+        {
+            openssl::DigestTypes::SHA1,
+            {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+            {'s', 'a', 'l', 't'},
+            2,
+            20,
+            "ea6c014dc72d6f8ccd1ed92ace1d41f0d8de8957"
+        },
+        {
+            openssl::DigestTypes::SHA1,
+            {'p', 'a', 's', 's', '\0', 'w', 'o', 'r', 'd'},
+            {'s', 'a', '\0', 'l', 't'},
+            4096,
+            16,
+            "56fa6aa75548099dcc37d7f03425e0c3"
+        },
+        {
+            openssl::DigestTypes::SHA512,
+            {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+            {'s', 'a', 'l', 't'},
+            1,
+            64,
+            "867f70cf1ade02cff3752599a3a53dc4af34c7a669815ae5d513554e1c8cf252c02d470a285a0501bad999bfe943c08f050235d7d68b"
+            "1da55e63f73b60a57fce"
+        },
+        {
+            openssl::DigestTypes::SHA512,
+            {'p', 'a', 's', 's', 'w', 'o', 'r', 'd'},
+            {'s', 'a', 'l', 't'},
+            4096,
+            64,
+            "d197b1b33db0143e018b12f3d1d1479e6cdebdcc97c5c0f87f6902e072f457b5143f30602641b3d55cd335988cb36b84376060ecd532"
+            "e039b742a239434af2d5"
+        }
+
+    };
+    return testData;
+}
+
+class Pbkdf2Tests : public testing::TestWithParam<inputData> {};
+INSTANTIATE_TEST_CASE_P(pbkdf2,
+                        Pbkdf2Tests,
+                        testing::ValuesIn(prepareTestDataForPbkdf2Tests()));
+
+TEST_P(Pbkdf2Tests, tests)
+{
+    testPbkfd2(GetParam());
+}
+
+
+
+
+
+
```

Comment:
Let's be a bit more swabian with the whitespace between those two functions?

Reply:
Sure. :)
Just thought they would help to differentiate between the PBKDF2 and X9.63 tests.